Komodor Addon Pattern

ronahk opened the pull request:
*Description of changes:*
New pattern for the [Komodor addon](https://github.com/komodorio/komodor-eks-blueprints-addon).


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

shapirov103 reviewed (commented):
@ronahk thank you for your contribution!

Please make sure your add-on submission is compliant with the following instructions:

- The add-on must be implemented according to the extensibility guidelines which we shared previously: https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/

- In particular sections for "Public Extensions" and "Partner Extensions": https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/#partner-extensions

- **Important**: Sensitive data must be passed according to the following section of the guide: https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/#passing-secrets-to-add-ons (done for security reasons)

In addition to the above, if your add-on requires credentials, then we will need them in order to test, unless the documentation that you submit to the [Quickstart repository](https://github.com/aws-quickstart/cdk-eks-blueprints) outlines a way to get free/evaluation access.

I will expect a PR against the Quickstart repo, with proper documentation, I will then will be able to test it out.

youngjeong46 reviewed (changes requested):
Hi @ronahk, thanks for submitting the pattern (and the reference doc on our quickstart repo). I notice that you provide ways to retrieve the API key using Secrets Manager Secret.

Implementation of that should follow the following section of the guide as @shapirov103 mentioned: https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/#passing-secrets-to-add-ons (done for security reasons)

Please make changes and resubmit when you can. Thank you!

ronahk commented:
Thanks for reviewing! @shapirov103 @youngjeong46 
I removed the option to retrieve the API key using Secrets Manager Secret (decided we don't need it) and released a new version of the addon which is now updated in this PR.
I also submitted a PR for documentation in the Quickstart repository, as requested. See [here](https://github.com/aws-quickstart/cdk-eks-blueprints/pull/593)
Hopefully this is ok. 
Thanks so much!

shapirov103 commented on lib/komodor-construct/index.ts:
```diff
@@ -0,0 +1,25 @@
+        const addOns: Array<blueprints.ClusterAddOn> = [
+            new KomodorAddOn({
+                clusterName: stackId,
+                apiKey: "<your-api-key>", //replace with your API key
```
API key appears to be a sensitive piece of information (similar to credentials) which is passed using non-secure approach. 
The CDK pattern will produce a CFN stack in the target account. Anyone who has access to the CFN stack will be able to look it up. 
We still have a path to approval, however, customers should be aware that this way of passing the security credential is insecure, and not recommended for production use cases. 

 If you decide to move forward with this implementation, then let's call out that this approach is not secure in the pattern (a comment) and in your documentation https://github.com/komodorio/komodor-eks-blueprints-addon. 

Alternatively, you can rework the implementation of the add-on to use CSI Secret Store driver.

Please let me know if you would like to discuss it in more details, come up with a secure implementation.

ronahk replied:
Hi @shapirov103 Can we please discuss it in more detail? 
Is there any addon that currently uses the CSI Secret Store driver?

shapirov103 reviewed (commented):
Added a comment, please review.

youngjeong46 reviewed (commented):
@ronahk just a gentle reminder to address the comment from @shapirov103.

shapirov103 commented:
@ronahk apologies for the late response, I generally get a flood of GitHub notifications, so must have missed it. I am happy to discuss the approach leveraging CSI Secret Store driver and AWS provider for it.  The approach is outlined [here](https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/#passing-secrets-to-add-ons) and example is available [here](https://github.com/shapirov103/eks-blueprints-extension).

And to answer your other question, some of the current production-ready add-ons support CSI Secret Store to comply with the security best practice, e.g New Relic.

youngjeong46 commented:
@ronahk We still see the secrets management as an issue on this one that needs to be resolved. As @shapirov103 mentioned above, using CSI Secret Store driver is the secured approach to ensure API Keys don't get exposed (which in the current implementation, will be). 

The approach to implementing is shown [here](https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/#passing-secrets-to-add-ons), with one of our partners having implemented [here](https://github.com/newrelic-experimental/newrelic-eks-blueprints-addon/blob/14bbe148c530a4f8c51b09162647418b73d42489/src/index.ts#L309). 

If you choose to forgo this approach, we need to document to make sure that the users are aware of security risks. That should be done on your addon repository: https://github.com/komodorio/komodor-eks-blueprints-addon

ronahk commented:
> here
Hi @youngjeong46 thanks for you comment! I understand that you see it as an issue. i'm not sure that you saw what i wrote to @shapirov103 but our case is exactly the same as Snyk (which didn't use the CSI Secret Store driver approach). 

Our Helm chart currently doesn’t support additional volumes / volumeMounts. I tried the approach that New Relic used which didn't work and I would be able to implement it after we've added support for volumes. Also the same in our case, the apiKey is “secret-ish” in the sense that it’s not a password, but an identifier for where to send data. We also opted to mark our addon as Beta for the time being, this is one of the reasons why.
If it's ok with you, I would like to continue with this implementation and improve it in the future. I would be happy to add a note about the apiKey in the README.

If it's not ok with you than I'll document to make sure that the users are aware of security risks, as you said.

shapirov103 reviewed (commented):
@ronahk if the add-on is marked as beta, then you can use less secure approaches, however, this decision may affect adoption and ultimately ROI. 

1. An alternative to the CSI Secret Store for production ready workloads is ExternalSecretsAddOn. That one can just create a secret in the target namespace, so if you can just access secrets or wire it to be passed as env variables, that would acceptable. 
2. NewRelic approach also required troubleshooting and we assisted with it. We can do the same for you if you would like. 
3. If you go with the current state and beta-level release, then we will accept this PR as is provided it passes technical validation.

@youngjeong46 ^^

ronahk commented:
> @ronahk if the add-on is marked as beta, then you can use less secure approaches, however, this decision may affect adoption and ultimately ROI.
> 
> 1. An alternative to the CSI Secret Store for production ready workloads is ExternalSecretsAddOn. That one can just create a secret in the target namespace, so if you can just access secrets or wire it to be passed as env variables, that would be acceptable.
> 2. NewRelic approach also required troubleshooting and we assisted with it. We can do the same for you if you would like.
> 3. If you go with the current state and beta-level release, then we will accept this PR as is provided it passes technical validation.
> 
> @youngjeong46 ^^

Thanks for the informitive reply! I would be happy to continue with option number 3 for the moment and improve it later on. 
Please let me know if i anything else needs to be done on my side.

youngjeong46 commented:
@ronahk One thing we are missing is the `bin/komodor.ts` file that exposes the pattern itself. We recently made a small change on pattern deployment. Please see [here](https://github.com/aws-samples/cdk-eks-blueprints-patterns#new-patterns), specifically #3.

ronahk commented:
Done, please review

youngjeong46 reviewed (approved):
LGTM. Tech Validation passed.
